readers/heuristics/src/test/java/com/criteo/hadoop/garmadon/heuristics/HeuristicsResultDBTest.java

```diff
@@ -0,0 +1,160 @@
+
+        ResultSet rset = db.connection.createStatement().executeQuery("SELECT  * FROM garmadon_yarn_app_heuristic_result");
+        HashMap<String, String> result = new HashMap<>();
+        while (rset.next()) {
+            result.put(rset.getString(1), rset.getString(2));
+        }
+        assertEquals(appId, result.get("1"));
+
+        rset = db.connection.createStatement().executeQuery("SELECT  * FROM garmadon_yarn_app_heuristic_result_details");
+        result = new HashMap<>();
+        while (rset.next()) {
+            result.put(rset.getString(1), rset.getString(2));
+        }
+        assertEquals(detail, result.get("1"));
+    }
+
+    @Test
+    public void updateHeuristicHelp() throws SQLException {
+        String help = "<p>\nTest\n</p>";
+        HeuristicsResultDB db = new HeuristicsResultDB(dbConfiguration);
+
+        List<Heuristic> heuristics = new ArrayList<>();
+        Heuristic heuristic = mock(Heuristic.class);
+        doReturn(help).when(heuristic).getHelp();
+        heuristics.add(heuristic);
+        db.updateHeuristicHelp(heuristics);
+
+        ResultSet rset = db.connection.createStatement().executeQuery("SELECT  * FROM garmadon_heuristic_help");
+        HashMap<String, String> result = new HashMap<>();
+        while (rset.next()) {
+            result.put(rset.getString(1), rset.getString(2));
+        }
+        assertEquals(help, result.get(heuristic.getClass().getSimpleName()));
+    }
+
+    @Test
+    public void reconnectOnFailure() {
+        String appId = "appId";
+        String detail = "detail";
+        HeuristicsResultDB db = spy(new HeuristicsResultDB(dbConfiguration));
+        HeuristicResult heuristicResult = new HeuristicResult(appId, "attemptId", GCCause.class, 1, 1);
+        heuristicResult.addDetail(detail, "value", "This is a new detail");
+        new Thread(() -> {
+            mysqld.stop();
+            initDb();
+        }).start();
+        waitUntilDbStopped();
+        db.createHeuristicResult(heuristicResult);
+        verify(db, atLeast(2)).initConnectionAndStatement(dbConfiguration);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void throwRuntimeExceptionIfInsertFailed() {
+        String appId = "appId";
+        String detail = "detail";
+        HeuristicsResultDB db = new HeuristicsResultDB(dbConfiguration);
+        HeuristicResult heuristicResult = new HeuristicResult(appId, "attemptId", GCCause.class, 1, 1);
+        heuristicResult.addDetail(detail, "value", "This is a new detail");
+        new Thread(() -> {
+            mysqld.stop();
+        }).start();
+        waitUntilDbStopped();
+        db.createHeuristicResult(heuristicResult);
+    }
+
+    private void waitUntilDbStopped() {
+        while (true) {
+            try {
+                new ServerSocket(port).close();
+                break;
+            } catch (IOException e) {
+                try {
+                    Thread.sleep(1L);
```

Comment:
sleep(1000) ?

Reply:
Wanted to be sure that I start to try inserting data immediately after the db is topped to avoid having the db restarted before inserting. Otherwise won't see any call to retry.
But I can update it to a bigger value (10 or a 100)